File: backend-python/app/providers/aws/storage.py
from __future__ import annotations

import os
import re
from datetime import datetime, timezone
from typing import Optional

try:
    import boto3
    from botocore.exceptions import ClientError
except ImportError:  # pragma: no cover
    boto3 = None  # type: ignore
    ClientError = Exception  # type: ignore

from ..base import StorageProvider


def _s3_client():
    if boto3 is None:
        raise RuntimeError("boto3 is required for AWS storage. Install it with: pip install boto3")
    region = os.getenv("AWS_REGION", "ap-southeast-2")
    return boto3.client(
        "s3",
        region_name=region,
        endpoint_url="https://s3.{0}.amazonaws.com".format(region),
    )


def _sanitize_file_name(file_name: str) -> str:
    sanitized = re.sub(r"[^a-zA-Z0-9._-]", "-", file_name.strip())
    sanitized = re.sub(r"-+", "-", sanitized).strip("-")
    return sanitized[:120] or "upload.bin"


class AwsStorageProvider(StorageProvider):
    def build_upload_blob_name(self, tenant_id: str, document_id: str, file_name: str) -> str:
        date_folder = datetime.now(timezone.utc).strftime("%Y/%m/%d")
        return "{0}/{1}/{2}-{3}".format(
            tenant_id, date_folder, document_id, _sanitize_file_name(file_name)
        )

    def create_upload_url(
        self,
        blob_name: str,
        container_name: str,
        expiry_minutes: int,
        content_type: Optional[str],
    ) -> str:
        """Generate a pre-signed S3 PUT URL for direct client upload."""
        client = _s3_client()
        params: dict = {
            "Bucket": container_name,
            "Key": blob_name,
        }
        if content_type:
            params["ContentType"] = content_type
        return client.generate_presigned_url(
            "put_object",
            Params=params,
            ExpiresIn=expiry_minutes * 60,
        )

    def download_blob(self, container_name: str, blob_name: str) -> bytes:
        client = _s3_client()
        response = client.get_object(Bucket=container_name, Key=blob_name)
        return response["Body"].read()

    def get_blob_content_type(self, container_name: str, blob_name: str) -> Optional[str]:
        client = _s3_client()
        try:
            response = client.head_object(Bucket=container_name, Key=blob_name)
            return response.get("ContentType")
        except ClientError:
            return None

    def delete_blob(self, container_name: str, blob_name: str) -> None:
        client = _s3_client()
        client.delete_object(Bucket=container_name, Key=blob_name)

File: backend-nodejs/functions-ingestion/src/providers/aws/storage.ts
import {
  S3Client,
  GetObjectCommand,
  HeadObjectCommand,
  PutObjectCommand
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import type { StorageProvider } from "../base.js";

function sanitizeFileName(input: string): string {
  return input
    .trim()
    .replace(/[^a-zA-Z0-9._-]/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "")
    .slice(0, 120);
}

export class AwsStorageProvider implements StorageProvider {
  private region = process.env.AWS_REGION ?? "ap-southeast-2";
  private s3 = new S3Client({ region: this.region });

  buildUploadBlobName(
    tenantId: string,
    documentId: string,
    fileName: string
  ): string {
    const safeFileName = sanitizeFileName(fileName);
    const dateFolder = new Date().toISOString().slice(0, 10).replace(/-/g, "/");
    return `${tenantId}/${dateFolder}/${documentId}-${safeFileName}`;
  }

  async createUploadUrl(opts: {
    blobName: string;
    containerName: string;
    expiryMinutes: number;
    contentType?: string;
  }): Promise<string> {
    const bucket = process.env.S3_BUCKET_NAME ?? opts.containerName;
    const command = new PutObjectCommand({
      Bucket: bucket,
      Key: opts.blobName,
      ContentType: opts.contentType ?? "application/octet-stream"
    });
    return getSignedUrl(this.s3, command, {
      expiresIn: opts.expiryMinutes * 60
    });
  }

  async downloadBlob(containerName: string, blobName: string): Promise<Buffer> {
    const bucket = process.env.S3_BUCKET_NAME ?? containerName;
    const response = await this.s3.send(
      new GetObjectCommand({ Bucket: bucket, Key: blobName })
    );
    const stream = response.Body;
    if (!stream) throw new Error(`Empty body for blob: ${blobName}`);
    // collect stream
    const chunks: Uint8Array[] = [];
    // @ts-ignore – Node.js ReadableStream
    for await (const chunk of stream) {
      chunks.push(chunk);
    }
    return Buffer.concat(chunks);
  }

  async getBlobContentType(
    containerName: string,
    blobName: string
  ): Promise<string | undefined> {
    const bucket = process.env.S3_BUCKET_NAME ?? containerName;
    const response = await this.s3.send(
      new HeadObjectCommand({ Bucket: bucket, Key: blobName })
    );
    return response.ContentType;
  }
}
